### PR TITLE
mgr/dashboard_v2: Add `@AuthRequired`

### PR DESCRIPTION
**Pull request by sebastian-philipp**
... to the cluster controller and to the api root.

Signed-off-by: Sebastian Wagner <sebastian.wagner@suse.com>

### CODE REVIEW

File: src/pybind/mgr/dashboard_v2/module.py [hunk 1]
```diff
@@ -156,6 +156,7 @@ def handle_command(self, cmd):
     def notify(self, notify_type, notify_id):
         NotificationQueue.new_notification(notify_type, notify_id)
 
+    @AuthRequired()
```
Comment (rjfd) [inline src/pybind/mgr/dashboard_v2/module.py:166]: If you add this to `ApiRoot`, how are you allowed to make a `POST` request to `/api/auth` to login in the first place?

File: src/pybind/mgr/dashboard_v2/tools.py [hunk 2]
```diff
@@ -84,6 +80,11 @@ def json_error_page(status, message, traceback, version):
                            version=version))
 
 
+_cp_config_default = {
```
Comment (votdev) [inline src/pybind/mgr/dashboard_v2/tools.py:83]: Why not let it part of the BaseControllerMeta?

**Review by votdev — DISMISSED** [review]

File: src/pybind/mgr/dashboard_v2/tools.py [hunk 3]
```diff
@@ -106,10 +107,6 @@ class BaseController(six.with_metaclass(BaseControllerMeta, object)):
     """
     _mgr_module = None
 
-    _cp_config_default = {
```
Comment (rjfd) [inline src/pybind/mgr/dashboard_v2/tools.py:109]: Why are you removing this from the `BaseControllerMeta`?

**Review by rjfd — CHANGES_REQUESTED** [review]

File: src/pybind/mgr/dashboard_v2/tools.py (hunk 2 above)
Comment (sebastian-philipp) [reply src/pybind/mgr/dashboard_v2/tools.py:83]: `AuthRequired()` can now be used independent of `BaseController`. `ApiRoot` is not a controller

File: src/pybind/mgr/dashboard_v2/module.py (hunk 1 above)
Comment (sebastian-philipp) [reply src/pybind/mgr/dashboard_v2/module.py:166]: The tests did pass!

File: src/pybind/mgr/dashboard_v2/tools.py (hunk 3 above)
Comment (sebastian-philipp) [reply src/pybind/mgr/dashboard_v2/tools.py:109]: `AuthRequired()` can now be used independent of `BaseController`. `ApiRoot` is not a controller

**Comment by rjfd** [comment]
@sebastian-philipp what is the problem you are trying to solve with this? making all controllers to require authentication by default?

**Comment by sebastian-philipp** [comment]
For a future PR, we should make authentication-required the default.

**Review by p-se — APPROVED** [review]

**Comment by rjfd** [comment]
@sebastian-philipp I don't like the idea of decoupling the `@AuthRequired` decorator from the controller architecture that we designed. Having the restriction of only allowing the use of `@AuthRequired` in classes that extend the `BaseController` class, although it narrows the possible implementation solutions that a developer can think of, it also benefits to maintain a coherent code structure and eases the review process.

This PR does two things:
a) protects the `Cluster` controller from unauthorized access; and
b) protects the `/api` URL from unauthorized access

Regarding a) we don't need that controller for the initial PR, so that should be removed from this PR.
Regarding b) `/api` is not exposed as a controller (it is not provided by a class that extends the `BaseController`) and is part of the infrastructure itself. Since this an exceptional case, and I agree that we should protect it against unauthorized access, the solution that makes more sense is just to active authentication using the `_cp_config`  class variable like this:

```
diff --git a/src/pybind/mgr/dashboard_v2/module.py b/src/pybind/mgr/dashboard_v2/module.py
index 564840a607..d52d02b6c2 100644
--- a/src/pybind/mgr/dashboard_v2/module.py
+++ b/src/pybind/mgr/dashboard_v2/module.py
@@ -165,6 +165,11 @@ class Module(MgrModule):
 
     class ApiRoot(object):
 
+        _cp_config = {
+            'tools.sessions.on': True,
+            'tools.authenticate.on': True
+        }
+
         def __init__(self, mgrmod):
             self.ctrls = load_controllers(mgrmod)
             logger.debug('Loaded controllers: %s', self.ctrls)

```

This requires much less changes, does not change the semantics of `@AuthRequired`, and fixes the problem.

**Comment by sebastian-philipp** [comment]
I agree

**Pull request closed without merging**